upload_vehicle_data.py

```diff
@@ -0,0 +1,380 @@
+
+
+def delete_remote_directory(remote_user, remote_ip, remote_directory):
+    """Delete a directory on the remote machine."""
+    command = f'rm -rf {remote_directory}/temp'
+    run_ssh_command(remote_user, remote_ip, command)
+
+
+def delete_remote_directory_contents(remote_user, remote_ip, remote_directory):
+    """Delete the contents of a directory on the remote machine."""
+    command = f'rm -rf {remote_directory}/*'
+    run_ssh_command(remote_user, remote_ip, command)
+
+
+def compress_and_transfer_rosbag(
+    remote_user,
+    remote_ip,
+    rosbag_path,
+    remote_directory,
+    cloud_upload_directory,
+):
+    """Compress rosbag on remote machine and transfer it to cloud host."""
+    remote_temp_directory = f'{remote_directory}/temp'
+    # Check available disk space before compression
+    if not check_disk_space(
+        remote_user, remote_ip, remote_temp_directory, rosbag_path
+    ):
+        logging.error(f'Insufficient disk space for compressing {rosbag_path}')
+        return False
+
+    # Compress the rosbag on the remote machine
+    remote_compressed_path = os.path.join(
+        remote_temp_directory, os.path.basename(rosbag_path)
+    )
+    mcap_path = os.path.join(os.getenv('HOME'), 'mcap')
+    compress_cmd = (
+        f'{mcap_path} compress {rosbag_path} -o {remote_compressed_path}'
+    )
+    try:
+        run_ssh_command(remote_user, remote_ip, compress_cmd)
+        logging.info(
+            f'Compressed rosbag {rosbag_path} to'
+            f'{remote_compressed_path} on the remote machine.'
+        )
+    except subprocess.CalledProcessError as e:
+        logging.error(
+            f'Failed to compress rosbag {rosbag_path} on remote machine: {e}'
+        )
+        return False
+
+    # Transfer the compressed file to the cloud host
+    rsync_cmd = [
+        'rsync',
+        '-avz',
+        '--checksum',
+        f'{remote_user}@{remote_ip}:{remote_compressed_path}',
+        cloud_upload_directory,
+    ]
+    try:
+        subprocess.run(rsync_cmd, check=True)
+        logging.info(
+            f'Transferred compressed rosbag {remote_compressed_path}'
+            f'to the cloud host at {cloud_upload_directory}.'
+        )
+    except subprocess.CalledProcessError as e:
+        logging.error(
+            f'Failed to transfer compressed rosbag'
+            f'{remote_compressed_path} from remote machine: {e}. Retrying...'
+        )
+        try:
+            subprocess.run(rsync_cmd, check=True)
+            logging.info(
+                f'Successfully retried and transferred compressed rosbag '
+                f'{remote_compressed_path} to the cloud host'
+                f'at {cloud_upload_directory}.'
+            )
+        except subprocess.CalledProcessError as e:
+            logging.error(
+                f'Retry failed to transfer compressed rosbag'
+                f'{remote_compressed_path} from remote machine: {e}'
+            )
+        return False

```

Comment:
Rather than hard-coding the two uploading attempts, we can put this in a while loop that retry uploading N times:

```suggestion
    success = False
    while attempts < max_upload_attempts:
        try:
            subprocess.run(rsync_cmd, check=True)
            logging.info(
                f'Transferred compressed rosbag {remote_compressed_path} '
                f'to the cloud host at {cloud_upload_directory}.'
            )
            success = True
            break
        except subprocess.CalledProcessError as e:
            attempts += 1
            logging.error(
                f'Failed to transfer compressed rosbag {remote_compressed_path} '
                f'from remote machine: {e}. Attempt {attempts} of {max_attempts}. Retrying...'
            )

    if not success:
        logging.error(
            f'All {max_attempts} attempts to transfer compressed rosbag '
            f'{remote_compressed_path} from remote machine have failed.'
        )
        return False

```



If, for some reason, this fails after N attempts, we need to ensure we know the name of the rosbag so that we do not delete its original and compressed versions after returning false. In other words, we need to track which files were successfully uploaded and which were not. This way, `delete_remote_directory_contents` will only delete uncompressed bags that were successfully uploaded.

Please add `upload_attempts` YAML parameter and amend wherever is required.

Reply:
logic changed.